### PR TITLE
Add assertRender to check rendered virtual DOM

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** feature

The following has been addressed in the PR:

* [X] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This PR is part of a larger effort to better unit test widgets.  It mainly provides the functionality to assert actual versus expected Dojo virtual DOM.

For example, once you have a value from a `render()` function (a `DNode`) you would compare it with expectations something like this this:

 ```ts
assertRender(results, v('div', {
    classes: [ css.root ]
}, [ w(SubWidget, { open: true }) ]), 'the results should match expectations');
```

The render is compared deeply (in the sense that properties, widget constructors, and children are checked for equality) but functions, like event handlers are simply checked for presence.

### CODE REVIEW

**Comment by agubler** [comment]
Do you think it makes sense for `compare` to live somewhere more "core"?

File: package.json [hunk 1]
```diff
@@ -27,19 +27,29 @@
...
   },
   "peerDependencies": {
-    "intern": "^3.4.1"
+    "@dojo/core": "beta1",
+    "@dojo/has": "beta1",
+    "@dojo/i18n": "beta1",
+    "@dojo/shim": "beta1",
+    "@dojo/widget-core": "beta1",
+    "intern": "^3.4.3",
+    "maquette": "2.4.5"
```
Comment (agubler) [inline package.json]: I think that widget-core is pinned on `2.4.3`....

File: package.json [hunk 2]
```diff
@@ -27,19 +27,29 @@
...
     "@types/grunt": "0.4.*",
+    "@types/jsdom": "2.0.*",
     "@types/sinon": "^1.16.31",
     "grunt": "~1.0.1",
-    "grunt-dojo2": ">=2.0.0-beta.34",
-    "intern": "^3.4.1",
-    "tslint": "^3.15.0",
+    "grunt-dojo2": "beta1",
+    "intern": "^3.4.3",
+    "tslint": "^4.5.1",
```
Comment (agubler) [inline package.json]: grunt-dojo2 should provide this now.

File: src/support/AssertionError.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	prototype: AssertionError;
+}
+
+/**
+ * ### AssertionError
+ *
+ * An extension of the JavaScript `Error` constructor for
+ * assertion and validation scenarios.
+ *
+ * @param {String} message
```
Comment (agubler) [inline src/support/AssertionError.ts]: looks like message is optional too - also shouldn't need types in doc :)

File: src/support/compare.ts [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+ * A guard that determines if the value is a plain object.  A plain object is an object that has
+ * either no constructor (e.g. `Object.create(null)`) or has Object as its constructor.
+ *
+ * @param value The value to check
+ */
+function isPlainObject(value: any): value is Object {
+	return Boolean(
+		value &&
+		typeof value === 'object' &&
+		value !== null &&
```
Comment (agubler) [inline src/support/compare.ts]: do you need `value` and `value !== null`?

File: src/support/AssertionError.ts (hunk 3 above)
Comment (kitsonk) [reply src/support/AssertionError.ts]: I was copypasta from the original code, I will strip it

File: src/support/compare.ts (hunk 4 above)
Comment (kitsonk) [reply src/support/compare.ts]: yup, correct

**Comment by kitsonk** [comment]
> Do you think it makes sense for `compare` to live somewhere more "core"?

Yes, see dojo/core#233.  This is essentially that, but with a couple of fixes I found while using it.  The latest, which isn't reflected in that conversation is that @pottedmeat talked to @maier49 about what was in stores versus what was in that PR and they agreed that for dgrid that @pottedmeat would take the API in the PR and use it and that we would evolve it from there.

I am fine refreshing that PR with this, but I didn't want to continue to wait around for us to sort something out which will be a dev only package.  When/if we ever landed dojo/core#233 and published, it would be an easy migration.

**Comment by agubler** [comment]
> I am fine refreshing that PR with this, but I didn't want to continue to wait around for us to sort something out which will be a dev only package. When/if we ever landed dojo/core#233 and published, it would be an easy migration.

Okay we should refresh the outstanding PR, but I understand the reasoning around including the module here for now.

File: src/support/compare.ts [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+	);
+}
+
+/**
+ * A guard that determines if the value is a primative (including `null`), as these values are
+ * fine to just copy.
+ *
+ * @param value The value to check
+ */
+function isPrimative(value: any): value is (string | number | boolean | undefined | null) {
```
Comment (agubler) [inline src/support/compare.ts]: @kitsonk Should this be `isPrimitive`?

Comment (kitsonk) [reply src/support/compare.ts]: 👍 thanks

**Review by agubler — APPROVED** [review]
👍

**Pull request merged**